File: de.prob2.kernel/src/main/resources/ui/bunit/BUnit.js
BUnit = (function() {
    var extern = {}
    var session = Session()
    
    $(document).ready(function() {
    });

    function clearInput() {
        $("#content").replaceWith(session.render("/ui/bunit/content.html", {}))
    }

    function setStats(data) {
        $("#stats").replaceWith(session.render("/ui/bunit/stats.html", data))
    }

    function addSuite(data) {
        $("#content").append(
                session.render("/ui/bunit/suite.html", data))
    }

    function addTest(suite, data) {
        $("#"+suite+"-list").append(session.render("/ui/bunit/test.html", data))
        $("#"+data.id).click(function(e) {
            $("#errors").addClass("invisible")
        })
    }

    function suiteFail(suite) {
    	$("#"+suite).addClass("test-suite-fail")
    }

    function suiteError(suite) {
    	$("#"+suite).addClass("test-suite-error")
    }

    function testIgnore(id) {
        $("#"+id).addClass("test-ignored")
        $("#"+id).click(function(e) {
            $("#errors").addClass("invisible")
        })
    }

    function testFail(id, reason) {
        $("#"+id).addClass("test-fail")
        $("#"+id).click(function(e) {
            $("#errors").removeClass("invisible")
            $("#errors").empty()
            $("#errors").html(reason)           
        }) 
    }

    function testError(id, reason) {
        $("#"+id).addClass("test-error")
        $("#"+id).click(function(e) {
            $("#errors").removeClass("invisible")
            $("#errors").empty()
            $("#errors").html(reason)           
        })        
    }



    extern.clearInput = clearInput

    extern.client = ""
    extern.init = session.init
    extern.addSuite = addSuite
    extern.addTest = function(data) {
        addTest(data.suite, data)
    }
    extern.testIgnore = function(data) {
        addTest(data.suite, data)
        testIgnore(data.id)
    }
    extern.testFail = function(data) {
    	suiteFail(data.suite)
        testFail(data.test, data.reason)
    }
    extern.testError = function(data) {
    	suiteError(data.suite)
        testError(data.test, data.reason)
    }
    extern.setStats = setStats

    return extern;
}())

File: de.prob2.kernel/src/main/java/de/prob/web/views/BUnit.java
package de.prob.web.views;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import javax.servlet.AsyncContext;

import com.google.inject.Singleton;

import de.prob.web.AbstractSession;
import de.prob.web.WebUtils;

@Singleton
public class BUnit extends AbstractSession {

	public BUnit() {
		incrementalUpdate = false;
	}

	Set<String> suites = new HashSet<String>();

	@Override
	public String html(final String clientid,
			final Map<String, String[]> parameterMap) {
		return simpleRender(clientid, "ui/bunit/index.html");
	}

	public void reset() {
		submit(WebUtils.wrap("cmd", "BUnit.clearInput"));
		suites.clear();
	}

	public void startTest(final String suiteName, final String testName,
			final String testId) {
		if (!suites.contains(suiteName)) {
			submit(WebUtils.wrap("cmd", "BUnit.addSuite", "name", suiteName));
			suites.add(suiteName);
		}
		submit(WebUtils.wrap("cmd", "BUnit.addTest", "suite", suiteName,
				"name", testName, "id", testId));
	}

	public void testFailure(final String suiteName, final String testId,
			final String message) {
		submit(WebUtils.wrap("cmd", "BUnit.testFail", "suite", suiteName,
				"test", testId, "reason", message));
	}

	public void testError(final String suiteName, final String testId,
			final String message) {
		submit(WebUtils.wrap("cmd", "BUnit.testError", "suite", suiteName,
				"test", testId, "reason", message));
	}

	public void testIgnore(final String suiteName, final String testName,
			final String testId) {
		if (!suites.contains(suiteName)) {
			submit(WebUtils.wrap("cmd", "BUnit.addSuite", "name", suiteName));
			suites.add(suiteName);
		}
		submit(WebUtils.wrap("cmd", "BUnit.testIgnore", "suite", suiteName,
				"name", testName, "id", testId));
	}

	public void testRunFinished(final int runCount, final int failCount,
			final int ignoreCount) {
		submit(WebUtils.wrap("cmd", "BUnit.setStats", "success", runCount
				- failCount, "total", runCount, "ignored", ignoreCount));
	}

	@Override
	public void reload(final String client, final int lastinfo,
			final AsyncContext context) {
		sendInitMessage(context);
		reset();
	}
}
